### ISSUE TITLE
Optimization of Ranking Loss function

### ISSUE THREAD

**Issue by prithagupta** [gabzai/Ranking-Loss-SCA#1]
The current ranking loss takes 405 ms to execute and is not compatible with tensorflow 2.0

The following ranking loss is compatible with tensorflow 2.0 and takes 439 ms to execute for 20 classes is as follows:

```
def ranking_loss(alpha_value=10, nb_class=256):
    def loss(y_true, y_pred):
        alpha = K.constant(alpha_value, dtype='float32')
        # Batch_size initialization
        y_true = to_categorical(y_true, num_classes=nb_class)

        y_true_int = K.cast(y_true, dtype='int32')
        batch_s = K.cast(K.shape(y_true_int)[0], dtype='int32')

        # Indexing the training set (range_value = (?,))
        range_value = K.arange(0, batch_s, dtype='int64')

        # Get rank and scores associated with the secret key (rank_sk = (?,))
        # values_topk_logits = shape(?, nb_class) ; indices_topk_logits = shape(?, nb_class)
        values_topk_logits, indices_topk_logits = tf.nn.top_k(y_pred, k=nb_class, sorted=True)
        rank_sk = tf.where(tf.equal(K.cast(indices_topk_logits, dtype='int64'),
                                    tf.reshape(K.argmax(y_true_int), [tf.shape(K.argmax(y_true_int))[0], 1])))[:, 1] + 1
                                    # Index of the correct output among all the hypotheses (shape(?,))
        score_sk = tf.gather_nd(values_topk_logits, K.concatenate(
            [tf.reshape(range_value, [tf.shape(values_topk_logits)[0], 1]),
             tf.reshape(rank_sk - 1, [tf.shape(rank_sk)[0], 1])]))  # Score of the secret key (shape(?,))
        # Ranking Loss Initialization
        loss_rank = 0
        #print('score_sk', score_sk)
        #print('rank_sk', rank_sk)
        for i in range(nb_class):
            # Score for each key hypothesis (s_i_shape=(?,))
            s_i = tf.gather_nd(values_topk_logits, K.concatenate(
                [tf.reshape(range_value, [tf.shape(values_topk_logits)[0], 1]),
                 i * tf.ones([tf.shape(values_topk_logits)[0], 1], dtype='int64')]))
            # Indicator function identifying when (i == secret key)
            indicator_function = tf.ones(batch_s) - (
                    K.cast(K.equal(rank_sk - 1, i), dtype='float32') * tf.ones(batch_s))
            # Logistic loss computation
            logistic_loss = K.log(1 + K.exp(-1 * alpha * (score_sk - s_i))) / K.log(2.0)
            # Ranking Loss computation
            loss_rank = tf.reduce_sum((indicator_function * logistic_loss)) + loss_rank

        return loss_rank / (K.cast(batch_s, dtype='float32'))

    return loss
```

This ranking loss function can be optimized as follows takes 23 ms to execute
```
def ranking_loss_optimized(alpha_value=10):
    def loss(y_true, y_pred):
        alpha = K.constant(alpha_value, dtype='float32')
        #y_true = tf.cast(tf.math.argmax(y_true, axis=1), dtype=tf.int32)
        y_true_int = K.cast(y_true, dtype='int32')
        values_topk_logits, indices_topk_logits = tf.nn.top_k(y_pred, sorted=True)
        batch_s = K.cast(K.shape(y_true_int)[0], dtype='int32')
        batch_range = K.arange(0, batch_s, dtype='int32')

        # Get rank and scores associated with the secret key (rank_sk = (?,))
        # values_topk_logits = shape(?, nb_class) ; indices_topk_logits = shape(?, nb_class)
        indices = tf.concat([batch_range[:, None], y_true_int[:, None]], 1)
        score_sk = tf.gather_nd(y_pred, indices)
        # Ranking Loss sk*-sk
        loss_rank_total = score_sk[:, None]- values_topk_logits
        # Ranking Logistic Loss sum_{k \in K/k*} log2(1 + exp(sk*-sk))
        logistic_loss = K.log(1 + K.exp(-1 * alpha * (loss_rank_total))) / K.log(2.0)
        loss_rank = tf.reduce_sum(logistic_loss, axis=1) - tf.constant(1.)
        return tf.reduce_mean(loss_rank)

    return loss
```

To test them please find this [notebook](https://jovian.ai/prithagupta/test-ranking-loss)
I would like to request to change the current loss function which is compatible with the latest tensorflow.

**Comment by prithagupta** [comment]
In order to use the loss function,

```
scores = Dense(num_classes, activation=None, name='scores')(x)
scoring_model = Model(inputs=trace_input, outputs=scores, name='cnn_baseline_scorer')​
scoring_model.compile(loss=ranking_loss_optimized(alpha_value=1), optimizer=RMSprop(learning_rate=0.00001), metrics=['accuracy'])
model.compile(loss=ranking_loss_optimized(alpha_value=1), optimizer=RMSprop(learning_rate=0.00001), metrics=['accuracy'])
scoring_model.fit(X,y)
```

This way, one can use the actual scores to evaluate the loss function and not the softmax probability scores.

**Comment by gabzai** [comment]
Hello Pritha,
Thank you very much for your contribution. Indeed, your optimized implementation sounds useful to speed the execution up.
I will soonly update the repository.
